math/optimized_trig.h

```diff
@@ -0,0 +1,230 @@
+/*
+	Copyright 2016 - 2019 Benjamin Vedder	benjamin@vedder.se
+
+	This file is part of the VESC firmware.
+
+	The VESC firmware is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    The VESC firmware is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+    */
+
+#ifndef OPTIMIZED_TRIG_H_
+#define OPTIMIZED_TRIG_H_
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <math.h>
+#include <stdlib.h>
+
+/**
+ * @brief cosine6TaylorApproximation
+ * @param AQ coef, aq * 2^qN
+ * @param BQm coef, -bq * 2^qN
+ * @param CQ coef, cq * 2^qN
+ * @param DQm coef, -dq * 2^(16 + qN - 2*qX)
+ * @param X2 X^2 * 2^qX
+ * @param X2_16 X^2 * 2^16
+ * @return
+ */
+__attribute__((always_inline))
+inline static int32_t cosine6TaylorApproximation(const int32_t AQ, const int32_t BQm, const int32_t CQ, const int32_t DQm, const int32_t X2, const int32_t X2_16) {
+   // out = a - bx^2 + cx^4 - dx^6
+   //     = a - x^2*((b - x^2*(c - d*x^2))
+   //     = a - x^2*(((-bm) - x^2*(c - (-dm)*x^2))
+   //     = a + x^2*(bm + x^2*(c + dm*x^2))
+
+   int32_t output;
+#if 1  // TODO: This needs to be #ifdef'ed for platforms which do not have the appropriate THUMB-2 instructions
+   // ARM THUMB-2 assembly blob which takes advantage of the SMLAWt operator, which performs a multily and accumulate
+   asm ("SMLAWb %[Y3], %[X2], %[D], %[C]\n"  // (y1=c-d*x^2
+        "SMLAWb %[Y3], %[Y3], %[X2_16], %[B]\n"   // (y2=-(b-y1*x16^2))
```

Comment:
Perhaps you mean 2^16 in the comment? Also the same below in the C version.

Reply:
Oh I see, it is fine the way it stands. I wasn't reading the whole documentation. Perhaps some spaces around binary operators will help readability. We can close this thread!

Reply:
Maybe `(y2=-(b-y1*(x_16)^2))`?